python/ray/data/tests/test_optimize.py

```diff
@@ -29,23 +59,92 @@ def test_memory_sanity(shutdown_only):
-    info = ray.init(num_cpus=1, object_store_memory=1500e6)
-    ds = ray.data.range(10)
+class OnesSource(Datasource):
+    def prepare_read(
+        self,
+        parallelism: int,
+        n_per_block: int,
+    ) -> List[ReadTask]:
+        read_tasks: List[ReadTask] = []
+        meta = BlockMetadata(
+            num_rows=1,
+            size_bytes=n_per_block,
+            schema=None,
+            input_files=None,
+            exec_stats=None,
+        )
+
+        for _ in range(parallelism):
+            read_tasks.append(
+                ReadTask(lambda: [[np.ones(n_per_block, dtype=np.uint8)]], meta)
+            )
+        return read_tasks
+
+
+@pytest.mark.parametrize("lazy_input", [True, False])
+def test_memory_release_pipeline(shutdown_only, lazy_input):
+    context = DatasetContext.get_current()
+    # Disable stage fusion so we can keep reads and maps from being fused together,
+    # since we're trying to test multi-stage memory releasing here.
+    context.optimize_fuse_stages = False
+    # This object store allocation can hold at most 1 copy of the transformed dataset.
+    if lazy_input:
+        object_store_memory = 3000e6
+    else:
+        object_store_memory = 3000e6
+
+    n = 10
+    info = ray.init(num_cpus=n, object_store_memory=object_store_memory)
+    if lazy_input:
+        ds = ray.data.read_datasource(
+            OnesSource(),
+            parallelism=n,
+            n_per_block=100 * 1024 * 1024,
+        )
+    else:
+        ds = ray.data.from_items(list(range(n)), parallelism=n)
+
+    # Create a single-window pipeline.
+    pipe = ds.window(blocks_per_window=n)
 
     # Round 1.
-    ds = ds.map(lambda x: np.ones(100 * 1024 * 1024, dtype=np.uint8))
-    meminfo = memory_summary(info.address_info["address"], stats_only=True)
-    assert "Spilled" not in meminfo, meminfo
-
-    # Round 2.
-    ds = ds.map(lambda x: np.ones(100 * 1024 * 1024, dtype=np.uint8))
+    def gen(x):
+        import time
+
+        # TODO(Clark): Remove this sleep once we have fixed memory pressure handling.
+        time.sleep(2)
+        if isinstance(x, np.ndarray):
+            return x
+        else:
+            return np.ones(100 * 1024 * 1024, dtype=np.uint8)
+
+    pipe = pipe.map(gen)
+
+    def inc(x):
+        import time
+
+        # TODO(Clark): Remove this sleep once we have fixed memory pressure handling.
+        time.sleep(2)
+        return x + 1
+
+    num_rounds = 10
+    for _ in range(num_rounds):
+        pipe = pipe.map(inc)
+
+    for block in pipe.iter_batches(batch_size=None):
+        for arr in block:
+            np.testing.assert_equal(
+                arr,
+                np.ones(100 * 1024 * 1024, dtype=np.uint8) + num_rounds,
+            )
     meminfo = memory_summary(info["address"], stats_only=True)
-    # TODO(ekl) we can add this back once we clear inputs on eager exec as well.
-    # assert "Spilled" not in meminfo, meminfo
+    assert "Spilled" not in meminfo, meminfo
```

Comment:
This test might prove to be flaky due to the memory reclaiming issues within Ray, so we may need to put this in a retry loop similar to `test_memory_release_lazy_shuffle`.

Reply:
Ack. Even 2s not enough to escape the reclaiming uncertainty?

Reply:
1s was a bit flaky, didn't see 2s fail while testing locally, but who knows in CI. For context, we use an "OOM grace period" of [2 seconds](https://github.com/ray-project/ray/blob/9f88031d4fff7547c884ec304766acd41d73989f/src/ray/common/ray_config_def.h#L428-L430) in core Ray, where we let reference counting + garbage collection and object spilling free up space, so I'd expect this to be ~enough.